etcdctl: support exec watch in v3

gyuho opened the pull request:
Fix https://github.com/coreos/etcd/issues/8814.

xiang90 commented:
/cc @jpbetz can you take a look?

jpbetz commented:
Sure, I’ll have a look

vikin91 commented:
I'd love to see this released. This blocks migration of APIv2 to APIv3.

Are there any short-term hacks for 3.2.10 or 3.2.11  to run a command on a change of watched value?

gyuho commented:
@vikin91 We plan to release this in 3.3 (release candidate will be out this or early next year).
Can you give this branch a try?

vikin91 commented:
Sure, I will try it soon and report my experience.

vikin91 commented:
@gyuho , I tried the code and it works good. However, there is one difference between APIv2 and APIv3. In APIv2, the `watch-exec` command set two (or maybe more?) env variables `ETCD_WATCH_KEY` and `ETCD_WATCH_VALUE` that can contained respectively key and value of the watched path in etcd. In APIv3 these variables are not being set. It would be great if APIv3 could keep the same behaviour to make the migration as smooth as possible.

gyuho commented:
@vikin91 Thanks for trying out!

I would like to keep it as simple as possible for now--environment variable support for watch key/value can be easily added later.

ref.

https://github.com/coreos/etcd/blob/b48cf77abb78755c8654eb8da61bc90a32802e7d/etcdctl/ctlv2/command/exec_watch_command.go#L123-L129

jpbetz commented on etcdctl/ctlv3/command/watch_command.go:
```diff
@@ -92,30 +99,55 @@ func watchInteractiveFunc(cmd *cobra.Command, args []string) {
-		return nil, fmt.Errorf("bad number of arguments")
+func getWatchChan(c *clientv3.Client, args []string, interactive bool) (clientv3.WatchChan, []string, error) {
+	if interactive {
+		args = args[1:] // args[0]=="watch"
```
Put the assertion in this comment in to code, e.g.: `if args[0] != "watch { panic("args[0] must be 'watch' for interactive calls") }` ?

jpbetz commented on etcdctl/ctlv3/command/watch_command.go:
```diff
@@ -92,30 +99,55 @@ func watchInteractiveFunc(cmd *cobra.Command, args []string) {
+			}
+		}
+		osArgs := os.Args[idx+1:]
+		execArgs, err = findExecArgs(osArgs)
```
Avoid repeating the `findExecArgs` call by restructuring this as:

```
if !interactive {
  ...
  args := osArgs
}
execArgs, err = findExecArgs(args)
```
?

Also, should findExecArgs return both the args before `--` as well as the args after `--` so that the subsequent code can deal with them as cleanly separated lists? I'm concerned that checks like `len(args) == 2 ` below might not make sense if we don't cleanly separate the two args lists.

jpbetz commented on etcdctl/ctlv3/command/watch_command.go:
```diff
@@ -125,14 +157,54 @@ func getWatchChan(c *clientv3.Client, args []string) (clientv3.WatchChan, error)
+		return nil, nil
+	}
+
+	if idx == len(args)-1 {
```
Let's split the args and execArgs into two distinct lists here and then simplify all following, logic, e.g.:

```
func findExecArgs(args []string) ([]string, error) {

foundSep, idx := false, 0
 for idx = range args {
 	if args[idx] == "--" && idx != 0 {
 		foundSep = true
 		break
 	}
 }
 if !foundSep {
 	return args, nil, nil
 }
watchArgs, execArgs := args[:idx], args[idx+1:]

// check for various invalid states
...

return watchArgs, execArgs, nil

}
```

gyuho commented:
@jpbetz PTAL. Refactored `parseWatchArgs` function and addressed other feedback as well--logic got a bit more complicated, so I added more test cases. Everything else is the same.

Thanks!

debovema commented:
As reported by @vikin91, the feature is not usable without the environment variables (ETCD_WATCH_KEY, ETCD_WATCH_VALUE...) being set.
Aside from that it works with v3.3.0-rc.1
